IPython/utils/_process_posix.py

```diff
@@ -59,11 +61,18 @@ class ProcessHandler(object):
 
     @property
     def sh(self):
-        if self._sh is None:        
-            self._sh = pexpect.which('sh')
-            if self._sh is None:
-                raise OSError('"sh" shell not found')
-        
+        self._sh = os.environ['SHELL']
+        # ensure we got a clean shell from our environment
+        with open('/etc/shells', 'r') as s:
+            for line in s:
+                if re.match(r'(/[^/]+)+$', line):
+                    line = line.rstrip()
+                    if line == self._sh:
+                        return self._sh
+        # if shell not clean, defaults to /bin/sh
+        self._sh = pexpect.which('sh')
```

Comment:
All of the setting of `self._sh` logic should still be indented - we only want to set `self._sh` once, as before, not on every use of `.sh`.  It should only happen the first time this property is accessed, so please restore the `if self._sh is None:` above